Add CustomDebugDescription conformance to AnyKeyPath

benpious opened the pull request:
Implementation of [this pitch](https://forums.swift.org/t/pitch-add-customdebugdescription-conformance-to-anykeypath/58705), adds a conformance to `CustomDebugDescription` to `AnyKeyPath`. This will need an SE before it's merged. 

I haven't modified api-digester yet, which I think I need to do...? I intended to see what the CI says first.

benpious committed:
initial

benpious committed:
it works

demangling mostly works

fix dots

printing works

add tests

add conformance to AnyKeyPath

implement SPI

subscripts fully work

comments

use cross platform image inspection

remove unnecessary comment

fix

fix issues

add conditional conformance

add types

try to fix the api-digester test

cr feedback: move impls behind flag, remove addChain(), switch statement, fallthrough instead of if-elses, move import

cr feedback: refactor switch statement

fix #ifdef

reindent, cr feedback: removes manual memory management

fix missing whitespace

fix typo

fix indentation issues

switch to regexes

checks should test in on all platforms

print types in subscripts

add test for empty subscript

Update test/api-digester/stability-stdlib-abi-without-asserts.test

Co-authored-by: Xiaodi Wu <13952+xwu@users.noreply.github.com>

add commas

fix failing test

fix stdlib annotation

cr feedback: remove global, refactor ifdef

cr feedback: switch back to manual memory management

switch to 5.8 macro

add new weakly linked functions to the allowlist

fix one more failing test

more cr feedback

more cr feedback

benpious commented:
SE pull request: https://github.com/apple/swift-evolution/pull/1724

@jckarter tagging you here per the "how to contribute" doc in the swift compiler repo, as I assume you'll want to review this at some point

jckarter commented:
@swift-ci Please test

Azoy commented on stdlib/public/runtime/ReflectionMirror.cpp:
```diff
@@ -1092,3 +1092,28 @@ id swift_reflectionMirror_quickLookObject(OpaqueValue *value, const Metadata *T)
 
 #endif  // SWIFT_ENABLE_REFLECTION
+
+#include "ImageInspection.h"
```
Typically we add this to the top near the rest of the include statements.

Azoy commented on stdlib/public/runtime/ReflectionMirror.cpp:
```diff
@@ -1092,3 +1092,28 @@ id swift_reflectionMirror_quickLookObject(OpaqueValue *value, const Metadata *T)
+
+#include "ImageInspection.h"
+
+SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_API
```
These should probably be `SWIFT_RUNTIME_STDLIB_INTERNAL`.

benpious replied:
Thanks, I didn't know that was an option. I will try switching.

Azoy commented on stdlib/public/runtime/ReflectionMirror.cpp:
```diff
@@ -1092,3 +1092,28 @@ id swift_reflectionMirror_quickLookObject(OpaqueValue *value, const Metadata *T)
+
+#include "ImageInspection.h"
+
+SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_API
+const char *swift_keypath_dladdr(void *address) {
+    SymbolInfo info;
+    if (lookupSymbol(address, &info) == 0) {
+        return 0;
+    } else {
+        return info.symbolName.get();
+    }
+}
+
+SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_API
+const char *swift_keypathSourceString(char *name) {
+    size_t length = strlen(name);
+    std::string mangledName = keypathSourceString(name, length);
+    if (mangledName == "") {
+        return 0;
+    } else {
+        char *rtnValue = new char[mangledName.length() + 1];
+        strcpy(rtnValue, mangledName.c_str());
+        return rtnValue;
+    }
+}
```
These should be within the `SWIFT_ENABLE_REFLECTION` condition because we don't want them if reflection is turned off.

benpious replied:
I can change  it. Fwiw, this part can work without reflection metadata, that's why I left it in. 

My intention was that we would continue to make best effort attempts at producing something useful as long as the requisite data might be available, and that whether this part worked or not would be determined by the IFDEFs in ImageInspection.h. But I guess this still qualifies as reflection, even if it does not use the traditional mechanisms.

Azoy commented on lib/Demangling/NodePrinter.cpp:
```diff
@@ -3159,6 +3160,75 @@ void NodePrinter::printEntityType(NodePointer Entity, NodePointer type,
+                } else if (child->getKind() == Node::Kind::Identifier){
+                    return std::string(child->getText());
+                }
+                return invalid;
+            } else {
+                return invalid;
+            }
+        } else {
+            return invalid;
+        }
+    } else {
+        return invalid;
+    }
+    return invalid;
```
I think instead of having all these else statements, you can simply have the outermost one and remove the inner ones because they'll all escape outside.

Azoy commented on lib/Demangling/NodePrinter.cpp:
```diff
@@ -21,6 +21,7 @@
 #include <cassert>
 #include <cstdio>
 #include <cstdlib>
+#include <stack>
```
Unused?

benpious replied:
Yeah, from before I realized that swift-demangle could show me the exact structure of the names I was interested in and was writing graph search stuff to try to explore the structure. Thanks for catching this.

Azoy commented on lib/Demangling/NodePrinter.cpp:
```diff
@@ -3159,6 +3160,75 @@ void NodePrinter::printEntityType(NodePointer Entity, NodePointer type,
   }
 }
 
+std::string Demangle::keypathSourceString(const char *MangledName,
```
It might be a little more readable if inside this you used a switch to check if the node is subscript/variable:
```cpp
switch (child->getKind()) {
case Node::Kind::Subscript: {
  // stuff here
  break;
}

case Node::Kind::Variable: {
  // stuff here
  break;
}

default:
  break;
}
```

Azoy commented on stdlib/public/runtime/ReflectionMirror.cpp:
```diff
@@ -1092,3 +1092,28 @@ id swift_reflectionMirror_quickLookObject(OpaqueValue *value, const Metadata *T)
+    if (mangledName == "") {
+        return 0;
+    } else {
+        char *rtnValue = new char[mangledName.length() + 1];
```
I wonder if this function can take a closure that passes `mangledName.c_str()` here instead. (I'm somewhat nervous about using `new` and not having a related `delete`, even though it has `.deallocate()` on the Swift side).

benpious replied:
I'll look into this, but idk if that's possible. I'm guessing I'm limited to C language features in this context, and I don't think Swift allows captures in a convention(c) closure.

benpious replied:
With that said, I share your concerns about the allocation with the dealloc far away in code written in another language entirely.

benpious replied:
This should be resolved. You can't capture variables in a c convention closure, so I've returned a class instead using  `Unmanaged`. If there's a better way to guarantee that a `String` outlives the scope of the closure let me know and I'll switch.

Azoy commented on stdlib/public/core/KeyPath.swift:
```diff
@@ -3703,3 +3703,138 @@ internal func _instantiateKeyPathBuffer(
+                    }
+                    return _openExistential(base, do: project)
+                }
+                func addChain() {
```
I think you can replace this with a single `description.append(".")` at the beginning of the loop.

benpious replied:
My first version of this function was implemented like that, but unfortunately I don't think it works. Optionals require some special handling, otherwise we end up with stuff like `\Theme.background.?alpha`.

Azoy replied:
then I think you can it at the end of the loop and do a `description.removeLast()` once you're outside it.

benpious replied:
It wasn't quite that simple, but the suggestion gave me an idea for a different impl which I think addresses this.

Azoy commented on stdlib/public/core/KeyPath.swift:
```diff
@@ -3703,3 +3703,138 @@ internal func _instantiateKeyPathBuffer(
+                let (rawComponent, optNextType) = buffer.next()
+                let hasEnded = optNextType == nil
+                let nextType = optNextType ?? Self.valueType
+                func name(for offset: Int) -> String {
```
I think it'd be easier to read if you removed these inline functions and instead joined the similar cases in the switch and defined these implementation in those. For example:
```swift
case .struct(let offset), .class(let offset):
  #if SWIFT_ENABLE_REFLECTION
    ...
  #endif
  ...

case .get(_, let accessors, _),
     .mutatingGetSet(_, let accessors, _),
     .nonmutatingGetSet(_, let accessors, _):
  func project<Base>(_: Base.Type) -> String {
    ...
  }

  ...
```

Azoy commented on stdlib/public/core/KeyPath.swift:
```diff
@@ -3703,3 +3703,138 @@ internal func _instantiateKeyPathBuffer(
   }
 }
 
+@_silgen_name("swift_keypath_dladdr")
+internal func keypath_dladdr(_: UnsafeRawPointer) -> UnsafePointer<CChar>?
+
+@_silgen_name("swift_keypathSourceString")
+internal func demangle(
+    name: UnsafePointer<CChar>
+) -> UnsafePointer<CChar>?
+
+
+fileprivate func dynamicLibraryAddress<Base, Leaf>(
+    of pointer: ComputedAccessorsPtr,
+    _: Base.Type,
+    _ leaf: Leaf.Type
+) -> String {
+    let getter: ComputedAccessorsPtr.Getter<Base, Leaf> = pointer.getter()
+    let pointer = unsafeBitCast(getter, to: UnsafeRawPointer.self)
+    if let cString = keypath_dladdr(UnsafeRawPointer(pointer)),
+        let demangled = demangle(name: cString) {
+        defer {
+            demangled.deallocate()
+        }
+        return String(cString: demangled)
+    } else {
+        return "<computed \(pointer) (\(leaf))>"
+    }
+}
```
I think all of this and the actual implementation in `debugDescription` should be inside a `#if SWIFT_ENABLE_REFLECTION`.

benpious replied:
I just want to confirm, if `SWIFT_ENABLE_REFLECTION` is off, we should print `"(value cannot be printed without reflection)"`, right?

Azoy reviewed (commented):
Welcome! This looks pretty good! I just have a few nitpicks to hopefully make things a little easier to read. In general, the Swift project uses 2 space indentation instead of 4 both in the runtime and stdlib.

Azoy commented:
Oh, could it be possible to add the types in the subscript case? `\Foo.subscript(_: Int, bar: String)` reads very nicely (I don't remember if the type is fully realized in the demangling or if you can get a simple name for it).

benpious commented:
> Welcome! This looks pretty good! I just have a few nitpicks to hopefully make things a little easier to read. In general, the Swift project uses 2 space indentation instead of 4 both in the runtime and stdlib.

Thanks! 

I believe the issues with indentation should be addressed at this point. 

> Oh, could it be possible to add the types in the subscript case? `\Foo.subscript(_: Int, bar: String)` reads very nicely (I don't remember if the type is fully realized in the demangling or if you can get a simple name for it).

Yes! I didn't notice this earlier, but the type is there based on the output of `xcrun swift-demangle --tree-only`, and I'm sure I can find a way to get it out. Actually, I think we might be able to combine this with `_typeByName` to get the actual arguments that were passed to the subscript, which could read even better in the case of a simple integer, or a _lot_ worse if it's a very complex object being used.

benpious commented:
@Azoy I believe that all of the issues you raised have been addressed. 

I've been having issues building the full toolchain so I'm not sure if my changes to the api digester worked, but I do believe that the test failures on Windows and Linux should also be resolved.

xwu commented:
@swift-ci test

xwu commented on test/Interpreter/keypath.swift:
```diff
@@ -43,6 +73,20 @@ public func generic_class_constrained_keypath<U, V>(_ c: V) where V : GenericCon
+print(\Controller["abc"])
+// CHECK: \S.a
+print(\S.a)
+// CHECK: {{\\Controller\.subscript\(int: Int str: String _: Int\)|\\Controller\.<computed 0x.* \(Int\)>}}
```
Are there missing commas here, or does the output really lack commas separating the parameters (and if so, can they be put in)?

benpious replied:
No commas. Should be easy to add. 

As you may have noticed, in the original impl there were no type annotations, so I left out the commas in an attempt to duplicate the way unapplied functions are written in the swift compiler (no commas or type annotations, i.e. `self.myFunc(someArg: someOtherArg:)`). Tbh I didn't really think about how it looked now that we have annotations again -- I agree that this would be much better with them added.

xwu commented on test/Interpreter/keypath.swift:
```diff
@@ -43,6 +73,20 @@ public func generic_class_constrained_keypath<U, V>(_ c: V) where V : GenericCon
+print(\Controller[int: 0, str: "", 0])
+// CHECK: {{\\Controller\.thirdLabel|\\Controller\.<computed 0x.* \(Optional<MyLabel>\)>}}
+print(\Controller.thirdLabel)
+// CHECK: {{\\Controller\.subscript\(\)|\\Controller\.<computed 0x.* \(Optional<Int>\)>}}
```
Linux CI failure:

```
Input was:
<<<<<<
          1: label 
          2: \GenericController<Int>.label 
          3: label 
          4: \Controller.<computed 0x000055dafc924670 (Optional<MyLabel>)>!.<computed 0x000055dafc924720 (String)> 
          5: \Controller.<computed 0x000055dafc9248e0 (String)> 
          6: \S.a 
          7: \Controller.<computed 0x000055dafc924c00 (Int)> 
          8: \Controller.<computed 0x000055dafc924dc0 (Optional<MyLabel>)> 
check:91                                                                  X error: no match found
          9: \Controller.<computed 0x000055dafc924e70 (Int)> 
check:91     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>>>>>>
```

It does look like your subscript is supposed to return `Int`, not `Optional<Int>`.

benpious replied:
I'll check into this later today. Not sure how this happened; I did test this on MacOS by turning off the happy-path for image lookup. But hopefully it will be a very simple fix.

xwu reviewed (changes requested):
One of the Linux CI failures appears unrelated, but the other one is real.

xwu commented on test/api-digester/stability-stdlib-abi-without-asserts.test:
```diff
@@ -57,6 +57,7 @@ Func MutableCollection.moveSubranges(_:to:) has been removed
 Func MutableCollection.removeSubranges(_:) has been removed
 Func RangeReplaceableCollection.removeSubranges(_:) has been removed
 Struct AnyHashable has added a conformance to an existing protocol _HasCustomAnyHashableRepresentation
+Class AnyKeyPath has added a conformance to an existing protocol CustomDebugStringConvertible
```
```suggestion
Class AnyKeyPath has added a conformance to an existing protocol CustomDebugStringConvertible
Class KeyPath has added a conformance to an existing protocol CustomDebugStringConvertible
Class PartialKeyPath has added a conformance to an existing protocol CustomDebugStringConvertible
Class ReferenceWritableKeyPath has added a conformance to an existing protocol CustomDebugStringConvertible
Class WritableKeyPath has added a conformance to an existing protocol CustomDebugStringConvertible
```

benpious commented:
> One of the Linux CI failures appears unrelated, but the other one is real.

The Windows one also seemed like an infra issue to me when I saw it last night. For my context, is it common for the CI to be flaky?

benpious commented:
@xwu I believe I've addressed all of your comments. 

One thing I wanted to flag: currently, the availability of the new conformance is set to 5.7. Is there any chance that this makes it into 5.7, or should I change that to some other version?

xwu commented:
Thanks!
@swift-ci test

xwu commented:
> One thing I wanted to flag: currently, the availability of the new conformance is set to 5.7. Is there any chance that this makes it into 5.7, or should I change that to some other version?

Ah, thanks for flagging this: we are past making it into 5.7, but in any case the availability version for newly added things on the main branch should be `9999`, I believe.

benpious commented:
> > One thing I wanted to flag: currently, the availability of the new conformance is set to 5.7. Is there any chance that this makes it into 5.7, or should I change that to some other version?
> 
> Ah, thanks for flagging this: we are past making it into 5.7, but in any case the availability version for newly added things on the main branch should be `5.8` (which expands to OS versions `9999`). Let's let the CI run first though.

I will change to 5.8. 

Looks like CI failed the API digester test on MacOS. Not sure where to go from here in terms of debugging that. Linux looks like it might be the same (presumably spurious) error from before, but I'm not sure. Windows passed though!

xwu commented:
Wooo, yes I'm stumped by the API digester warning; it seems to me to be a false positive but I don't want to be hasty in suggesting that we add just add it to the list and call it aday; perhaps @Azoy can help here to verify.

xwu commented:
@swift-ci test Linux platform

Azoy commented on stdlib/public/runtime/ReflectionMirror.cpp:
```diff
@@ -1126,4 +1127,28 @@ id swift_reflectionMirror_quickLookObject(OpaqueValue *value, const Metadata *T)
+  if (mangledName == "") {
+    return 0;
+  } else {
+    return callback(mangledName.c_str());
```
Ok, I've went back and forth on this, but I think it'd just be easier to call `strdup` here and return that pointer from this function. Then on the Swift side you can call `String(cString:)` from that pointer and just `deallocate()` when you're done.

benpious commented on stdlib/public/core/OutputStream.swift:
```diff
@@ -638,3 +638,5 @@ internal struct _TeeStream<L: TextOutputStream, R: TextOutputStream>
   internal mutating func _unlock() { right._unlock(); left._unlock() }
 }
+
+internal let noReflectionMetadataErrorMessage = "(value cannot be printed without reflection)"
```
I will revert this. Can you explain/link to why the stdlib avoids globals? I would like to understand if the reason is exclusive to the stdlib or something that is applicable outside.

Azoy replied:
Globals have a place in the language, but I think we would reserve them for things like global caches if we had any of those (like the ones in the runtime). This is simple enough that I don't think it's worth having one.

benpious replied:
Ah, interesting. My guess was that it would be a perf issue related to initializing the let in a thread safe way. 

My goal with making this change was to prevent code duplication, but it's definitely not a particularly serious issue if this string gets copy pasted around a few times.

benpious commented on stdlib/public/core/KeyPath.swift:
```diff
@@ -3711,3 +3711,130 @@ internal func _instantiateKeyPathBuffer(
+
+#endif
+
+@available(SwiftStdLib 5.7, *)
```
Thanks! I was wondering why I saw warnings in the build log about it. 

Understood about the version number to use. Does this get grepped and replaced with the appropriate number when a release branch is cut?

Azoy replied:
Yep!

xwu replied:
@Azoy I think the macro expansion is actually `SwiftStdlib 5.8`, which expands out to `iOS 9999`, `macOS 9999`, etc.. There are no instances of `SwiftStdlib 9999` in the repository.

Azoy replied:
huh, I thought we had a 9999 macro. We should definitely add one later.

benpious replied:
The [documentation](https://github.com/apple/swift/blob/main/docs/StandardLibraryProgrammersManual.md) says to use 9999 "If we haven't defined a version number for the "next" Swift release yet." 

Based on the existence of a 5.8 constant in the codebase and some comments from Doug Gregor on the Swift forums it looks like 5.8 is "defined" as the next version to be released. Shall I change to 5.8 or keep 9999?

Azoy replied:
5.8 👍

benpious replied:
Changed to 5.8.

benpious commented on stdlib/public/core/KeyPath.swift:
```diff
@@ -3711,3 +3711,130 @@ internal func _instantiateKeyPathBuffer(
+  
+  @available(SwiftStdLib 5.7, *)
+  public var debugDescription: String {
+#if SWIFT_ENABLE_REFLECTION
```
I've changed this to not be nested in the function, matching [SourceCompatibilityShims.swift](https://github.com/apple/swift/blob/main/stdlib/public/Concurrency/SourceCompatibilityShims.swift#L79).

Azoy commented:
@swift-ci please test

xwu commented:
I haven't investigated deeply, but if it's the same Linux test falling again, and other CI builds on main are successful, then it's not a spurious error and probably needs some PR tweaking...

benpious commented:
> I haven't investigated deeply, but if it's the same Linux test falling again, and other CI builds on main are successful, then it's not a spurious error and probably needs some PR tweaking...

Yeah, it looks like it's this Pr. 

TL:DR I _think_ the right approach is to add these symbols to the allowlist of the test. But I would like to confirm with the author of the test (@compnerd). 

Long version:

My understanding is that the test lists all symbols in a binary, then diffs that output with a list of all strong symbols in the same binary. The goal is to ensure there are no weakly linked symbols at all. But there are two weakly linked symbols in the output: 
```
--- /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/stdlib/Output/symbol-visibility-linux.test-sh.tmp/swiftCore-all.txt	2022-09-05 01:04:57.351384015 +0000
+++ /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/stdlib/Output/symbol-visibility-linux.test-sh.tmp/swiftCore-no-weak.txt	2022-09-05 01:04:57.507385347 +0000
@@ -13089,8 +13089,6 @@
 000000000052bb90 D $syyXfWV
 000000000052bae0 D $syyYjrcWV
 000000000052ba88 D $syycWV
-000000000044b360 W _ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE17_M_realloc_insertIJS5_EEEvN9__gnu_cxx17__normal_iteratorIPS5_S7_EEDpOT_
-0000000000441690 W _ZStplIcSt11char_traitsIcESaIcEENSt7__cxx1112basic_stringIT_T0_T1_EERKS8_SA_
 000000000060b10b B __bss_start
 00000000003e6d60 T __gnu_f2h_ieee
 00000000003e6d20 T __gnu_h2f_ieee
```

Demangling these, it looks like we caused a new version of `std::operator+` and some massive, vector related template thing to be created.

Anyway, I _think_ what is happening is that I used these in the impls of some of the functions I wrote, without realizing that they're brand new templates that have never been used before. C++ is instantiating (is that the right word?) these templates, and using weak linking, for whatever reason. 

The test is filled with inverted matches which appear to be an allowlist, and based on the names it seems like they might also be templates. So the obvious, quick solution is to add the two functions there and move on. 

But I'd like to understand why this Pr causes this; nothing has changed in the allowlist since April. So it feels like there's something different (and potentially unidiomatic/wrong) about how I'm using C++ in this Pr.

compnerd commented:
:sigh: libstdc++.  Yes, these two instantiations should be okay to add to the allowlist.  The problem here is that the GNU C++ runtime decorates the template bodies such that an instantiation will be weak as it is in global namespace and attributed so as to coalesce the copies requiring weak ODR semantics giving them weak global linkage.  This is an artifact of the implementation and we cannot control that even with `-fvisibility=hidden -fvisibility-inlines-hidden` nor any internal macro.  This also changes based upon the version of the runtime in use as these are internal details.  The new code uses `+` on `std::string` and causes a `std::vector<std::string>` to potentially be resized when inserting (plausibly a `push_back`).

benpious commented:
> :sigh: libstdc++. Yes, these two instantiations should be okay to add to the allowlist. The problem here is that the GNU C++ runtime decorates the template bodies such that an instantiation will be weak as it is in global namespace and attributed so as to coalesce the copies requiring weak ODR semantics giving them weak global linkage. This is an artifact of the implementation and we cannot control that even with `-fvisibility=hidden -fvisibility-inlines-hidden` nor any internal macro. This also changes based upon the version of the runtime in use as these are internal details. The new code uses `+` on `std::string` and causes a `std::vector<std::string>` to potentially be resized when inserting (plausibly a `push_back`).

Thanks! I've added them to the allowlist.

xwu commented:
> :sigh: libstdc++.

Are there suitably equivalent LLVM types that can be used here instead?

benpious commented:
> > :sigh: libstdc++.
> 
> Are there suitably equivalent LLVM types that can be used here instead?

I don't believe so. I see two options in the llvm programmer's manual: 
- `StringRef`: this appears to be immutable, so not suitable for this use case 
- `Twine`: this initially looks promising as it's for "efficiently concatenating values together." But "its implementation relies on the ability to store pointers to temporary stack objects which may be deallocated at the end of a statement. Twines should only be used accepted as const references in arguments"

I think we could optimize the `matchSequenceOfKinds` function a little by switching to `llvm::smallVector`, because the size of the vector is known in advance. I can make that change if you like, but I'm pretty confident that it won't remove the need to modify the test.

xwu commented:
@swift-ci test

xwu commented:
> I think we could optimize the matchSequenceOfKinds function a little by switching to llvm::smallVector, because the size of the vector is known in advance. I can make that change if you like, but I'm pretty confident that it won't remove the need to modify the test.

I'll defer to you and other reviewers :)

xwu commented:
@swift-ci please test Linux platform

xwu commented:
@swift-ci please test macOS platform

xwu commented:
@swift-ci please test Windows platform

benpious commented:
Looks like CI is green. Any other changes needed to the Pr?

xwu commented on include/swift/Demangling/Demangle.h:
```diff
@@ -649,6 +649,11 @@ ManglingErrorOr<const char *> mangleNodeAsObjcCString(NodePointer node,
 
+/// Transforms a mangled key path accessor thunk helper
+/// into the identfier/subscript that would be used to invoke it in swift code.
+std::string keypathSourceString(const char *MangledName,
```
I _think_, based on a cursory look through other C++ sources, that we still spell it `keyPath` (note the capitalization) in compiler-internal contexts.

xwu commented on lib/Demangling/NodePrinter.cpp:
```diff
@@ -3211,6 +3212,139 @@ void NodePrinter::printEntityType(NodePointer Entity, NodePointer type,
+        case Node::Kind::Subscript: {
+          std::string subscriptText = "subscript(";
+          std::vector<std::string> argumentTypeNames;
+          // Multiple arguments case
```
End-of-line gremlin here.

xwu commented on stdlib/public/runtime/ReflectionMirror.cpp:
```diff
@@ -1126,4 +1127,25 @@ id swift_reflectionMirror_quickLookObject(OpaqueValue *value, const Metadata *T)
 #endif
 
+SWIFT_CC(swift) SWIFT_RUNTIME_STDLIB_INTERNAL
+const char *swift_keypath_dladdr(void *address) {
```
Same comment regarding `keypath` versus `keyPath` here and below.

xwu reviewed (commented):
Woooo, nice to see. Can you clang-format the .cpp files?

benpious committed:
fix invisible unicode

benpious commented:
> Woooo, nice to see. Can you clang-format the .cpp files?

Done.

xwu commented:
@swift-ci test

xwu commented:
@Azoy How's it look to you?

Azoy commented:
Thanks so much! Super excited this is in finally.

benpious commented:
> Thanks so much! Super excited this is in finally.

Me too! Thanks @Azoy @xwu and @compnerd for all the help getting this landed!
